web/api/report_server.thrift

```diff
@@ -467,8 +468,16 @@ union AnalysisInfoFilter {
   3: i64 reportId,
 }
 
+struct AnalysisInfoChecker {
+  1: optional bool enabled,  // If the checker was enabled during the analysis.
```

Comment:
I adhered to the [Thrift guidelines](https://diwakergupta.github.io/thrift-missing-guide/#defining-structs). If you don't say something is `optional`, it will be impossible to accurately deprecate and drop without breaking something. (The API file is full of `!DEPRECATED!` markers already because we did not care about this for a very long time.) In general, new API functions seemed to be often introduced as taking a `struct` member instead of multiple parameters. In returns values, changing from `map<X, Y>` to `map<X, struct<Y, Z>>` would be a horrible ABI break.

At a later point in time, maybe someone will decide that the enabled-ness of a checker is not enough, they want to obtain the exact checker options? That could be a reasonable addition here...

Or the notion of `enabled`-ness can change from a `bool` to something else, because we realise things could need to be transitively enabled...

Having this `optional` allows handling the lack of the value in the client codes without Thrift blowing up under us.

Reply:
I'm not sure if I agree with the first part. It would imply that every struct member and function parameter should be automatically optional. I prefer thinking of "optional" as a non-required thing, and not as a tool for providing backward compatibility. As far as I understand (please, correct me if I'm wrong) this optional provides backward compatibility during a potential API change by not requesting this field anymore. But if we decide to deprecate it, then we still need to leave this in the source code with a `!DEPRECATED!` label for the sake of old clients. Something makes me think that the number at the beginning of the line is used for supporting API changes.
My biggest technical concern about optional fields is that we are messaging to the clients that this field is really not necessary, and we need to check its existence at every single place where we use it. I'm not convinced it would be convenient, regarding that it really is a required field.

Reply:
> I prefer thinking of "optional" as a non-required thing, and not as a tool for providing backward compatibility.

The guidebook recommended by Thrift developers disagrees with this (emphasis mine):

> ⛔ **Required Is Forever**
> 
> You should be very careful about marking fields as required. If at some point you wish to stop writing or sending a required field, it will be problematic to change the field to an optional field — **old readers will consider messages without this field to be incomplete and may reject or drop them unintentionally.**

I based the changes on existing observed practice, e.g., https://github.com/Ericsson/codechecker/commit/4b4d2878f8fef57b5b334083470ee0034e825451, https://github.com/Ericsson/codechecker/commit/7b0dc2ede9e9ef214875f869260de55621362493 added new fields that are `optional`.

In https://diwakergupta.github.io/thrift-missing-guide/#_versioning_compatibility, it says that

> Any new fields that you add should be optional.

(Okay, technically, this patch here introduces the entire struct (all its one member), so this isn't just a "new field", but a whole new type...)

You are right that the number is used as the "slot" which Thrift internally uses to identify which piece of data belongs where. The *name* of the fields isn't serialised. This helps reduce the size of the transmitted packet. This is not directly related to API changes.

> But if we decide to deprecate it, then we still need to leave this in the source code with a `!DEPRECATED!` label for the sake of old clients.

This is only true if you are still sending that information. However, in many cases, we actually do not. For example, the run's deprecated `checkCommand` became `analysisInfo`, and when queried, an empty string (albeit a valid result...) is returned.

The problem is that with `bool` there isn't a good third option that could be the "default value once the field is deprecated", because both `false` and `true` mean something very specific in this context.

> Non-required fields can be removed, as long as the tag number is not used again in your updated message type (it may be better to rename the field instead, perhaps adding the prefix "OBSOLETE_", so that future users of your .thrift can’t accidentally reuse the number).